src/Compiler/Utilities/range.fs

```diff
@@ -581,3 +581,97 @@ module Range =
                     | None -> mkRange file (mkPos 1 0) (mkPos 1 80)
         with _ ->
             mkRange file (mkPos 1 0) (mkPos 1 80)
+
+module internal FileContent =
+    let private fileContentDict = ConcurrentDictionary<string, string>()
+
+    let update (fileName: string) (fileContent: string) =
+        fileContentDict.AddOrUpdate(fileName, (fun _ -> fileContent), (fun _ _ -> fileContent))
+        |> ignore
+
+    let private seperators = [| '\r'; '\n' |]
+
+    /// Find the index of the nearest line separator (\r or \r\n or \n) in the given string and offset.
+    let findLineEnd (input: string) lineStart =
+        if lineStart >= input.Length then
+            input.Length - 1
+        else
+            let idx = input.IndexOfAny(seperators, lineStart)
+
+            if idx = -1 then
+                input.Length - 1
+            elif input.[idx] = '\r' && idx + 1 < input.Length && input.[idx + 1] = '\n' then
+                idx + 1
+            else
+                idx
+
+    /// Get the substring of the given range.
+    /// This can retain the line seperators in the source string.
+    let substring (input: string) (range: range) =
```

Comment:
Considering the main usage of `fileContentDict` is to identify expressions based on a range:
1. I would for sure make it lazy to prevent allocating massive strings which no one will read
2. Wouldn't the algorithm be a lot simpler if that lazy storage worked with an array of lines instead of a single string? (writing that now, I do realize this would drop the ability to keep original line endings, wouldn't it?)